File: crazydog_ws/src/unitree_interfaces/src/unitree_pubsub.cpp
#include <chrono>
#include <iostream>
#include <thread>
#include <vector>
#include <cmath>
#include <rclcpp/rclcpp.hpp>
#include <unitree_msgs/msg/low_command.hpp>
#include <unitree_msgs/msg/low_state.hpp>
#include <unitree_msgs/msg/motor_command.hpp>
#include <unitree_msgs/msg/motor_state.hpp>
#include <std_msgs/msg/float32_multi_array.hpp>
#include <sensor_msgs/msg/joint_state.hpp>
#include <unistd.h>
#include "serialPort/SerialPort.h"
#include "unitreeMotor/unitreeMotor.h"

#define WHEEL_RADIUS 0.07  // m

using namespace std::chrono_literals;
using namespace unitree_msgs::msg;


class UnitreeMotor {
public:
    UnitreeMotor(int motor_id = 0, double MAX_degree = 0.0, double MIN_degree = 0.0, double initial_position_check = 0.0) {
        id = motor_id;
        cmd.motorType = MotorType::GO_M8010_6;
        data.motorType = MotorType::GO_M8010_6;

        initial_position_max = initial_position_check + 1.0;
        initial_position_min = initial_position_check - 1.0;
        initial_position = initial_position_check;

        max_position = initial_position_check + MAX_degree;
        min_position = initial_position_check + MIN_degree;
        max = MAX_degree;
        min = MIN_degree;

        cmd.id = motor_id;
    }

    // Public members for easy access.
    int id;
    MotorCmd cmd;
    MotorData data;
    double initial_position_max;
    double initial_position_min;
    double initial_position;
    double max_position;
    double min_position;
    double max;
    double min;
};

class UnitreeCommunication {
public:
    SerialPort serial;
    std::vector<UnitreeMotor> motors;

    UnitreeCommunication(const std::string &device_name) : serial(device_name) {}

    UnitreeMotor* createMotor(int motor_id, double max, double min, double initial_pos) {
        for (auto &motor : motors) {
            if (motor.id == motor_id) return &motor;
        }
        motors.emplace_back(motor_id, max, min, initial_pos);
        return &motors.back();
    }

    void enableAllMotors() {
        for (auto &motor : motors) {
            motor.cmd.mode = queryMotorMode(MotorType::GO_M8010_6, MotorMode::FOC);
            motor.cmd.q = 0;
            motor.cmd.dq = 0;
            motor.cmd.kp = 0;
            motor.cmd.kd = 0;
            motor.cmd.tau = 0;
            serial.sendRecv(&motor.cmd, &motor.data);
            std::this_thread::sleep_for(10ms);
        }
    }

    void positionForceVelocityCmd(int motor_id, double torque, double kp, double kd, double pos, double vel) {
        for (auto &motor : motors) {
            if (motor.id == motor_id) {
                motor.cmd.mode = queryMotorMode(MotorType::GO_M8010_6, MotorMode::FOC);
                motor.cmd.tau = torque;
                motor.cmd.kp = kp;
                motor.cmd.kd = kd;
                motor.cmd.q = pos;
                motor.cmd.dq = vel * queryGearRatio(MotorType::GO_M8010_6);
            }
        }
    }

    void motorSendRecv() {
        for (auto &motor : motors) {
            if (motor.data.q >= motor.min_position && motor.data.q <= motor.max_position) {
                serial.sendRecv(&motor.cmd, &motor.data);
            } else {
                RCLCPP_WARN(rclcpp::get_logger("unitree_communication"), 
                            "Motor %d out of constraint: [%f, %f]", motor.id, motor.min_position, motor.max_position);
                motor.cmd.q = 0;
                motor.cmd.dq = 0;
                motor.cmd.kp = 0;
                motor.cmd.kd = 0;
                motor.cmd.tau = 0;
                serial.sendRecv(&motor.cmd, &motor.data);
            }
        }
    }
};

class UnitreeInterface : public rclcpp::Node {
private:
    std::shared_ptr<UnitreeCommunication> unitree_left, unitree_right;
    rclcpp::Subscription<LowCommand>::SharedPtr command_sub;
    rclcpp::Publisher<LowState>::SharedPtr status_pub;
    rclcpp::Publisher<sensor_msgs::msg::JointState>::SharedPtr jointstate_pub;
    rclcpp::Subscription<std_msgs::msg::Float32MultiArray>::SharedPtr foc_status_sub;
    sensor_msgs::msg::JointState jointstate_msg;
    

public:
    UnitreeInterface() : Node("unitree_pubsub") {
        unitree_left = std::make_shared<UnitreeCommunication>("/dev/unitree-l");
        unitree_right = std::make_shared<UnitreeCommunication>("/dev/unitree-r");
        unitree_left->createMotor(1, 8.475, -5.364, 0.669);
        unitree_left->createMotor(2, 26.801, -1, 1.080);
        unitree_right->createMotor(4, 5.364, -8.475, 7.530); // Adjusted for initial position + 2*pi
        unitree_right->createMotor(5, 1, -26.801, 2.320);

        command_sub = this->create_subscription<LowCommand>(
            "unitree_command", 1, std::bind(&UnitreeInterface::commandCallback, this, std::placeholders::_1));
        status_pub = this->create_publisher<LowState>("unitree_status", 1);
        jointstate_pub = this->create_publisher<sensor_msgs::msg::JointState>("jointstate", 1);
        foc_status_sub = this->create_subscription<std_msgs::msg::Float32MultiArray>(
            "foc_msg", 1, std::bind(&UnitreeInterface::focStatusCallback, this, std::placeholders::_1));

        jointstate_msg.name = {"hip_l", "thigh_l", "calf_l", "hip_r", "thigh_r", "calf_r", "wheel_l", "wheel_r"};
        jointstate_msg.position.resize(8, 0.0);
        jointstate_msg.velocity.resize(8, 0.0);

        unitree_left->enableAllMotors();
        unitree_right->enableAllMotors();

        timer_ = this->create_wall_timer(1ms, std::bind(&UnitreeInterface::recvTimerCallback, this));
    }

    void commandCallback(const unitree_msgs::msg::LowCommand::SharedPtr msg) {
        for (size_t id = 0; id < msg->motor_cmd.size(); ++id) {
            int motor_number = static_cast<int>(id);
            double torque = msg->motor_cmd[id].tau;
            double kp = msg->motor_cmd[id].kp;
            double kd = msg->motor_cmd[id].kd;
            double position = msg->motor_cmd[id].q;
            double velocity = msg->motor_cmd[id].dq;

            // Send commands to both sets of motors
            unitree_left->positionForceVelocityCmd(motor_number, torque, kp, kd, position, velocity);
            unitree_right->positionForceVelocityCmd(motor_number, torque, kp, kd, position, velocity);
        }
    }

    void focStatusCallback(const std_msgs::msg::Float32MultiArray::SharedPtr msg) {
        if (msg->data[0] == 513) {
            jointstate_msg.position[6] = -msg->data[1];
            jointstate_msg.velocity[6] = -msg->data[2] * WHEEL_RADIUS * (2 * M_PI / 60);
        } else if (msg->data[0] == 514) {
            jointstate_msg.position[7] = msg->data[1];
            jointstate_msg.velocity[7] = msg->data[2] * WHEEL_RADIUS * (2 * M_PI / 60);
        }
    }

    void recvTimerCallback() {
        auto t0 = std::chrono::system_clock::now();
        unitree_left->motorSendRecv();
        unitree_right->motorSendRecv();
        auto t1 = std::chrono::system_clock::now();
        auto t0_seconds = std::chrono::duration_cast<std::chrono::duration<double>>(t0.time_since_epoch()).count();
        auto t1_seconds = std::chrono::duration_cast<std::chrono::duration<double>>(t1.time_since_epoch()).count();
        RCLCPP_INFO(this->get_logger(), "%f", t1_seconds-t0_seconds);

        auto msg_list = unitree_msgs::msg::LowState();
        // auto jointstate_msg = sensor_msgs::msg::JointState();

        // Set the header for joint state message
        jointstate_msg.header.stamp = this->get_clock()->now();
        jointstate_msg.header.frame_id = "";
        jointstate_msg.name = {"hip_l", "thigh_l", "calf_l", "hip_r", "thigh_r", "calf_r", "wheel_l", "wheel_r"};

        // Process motors from the first unitree interface
        for (const auto &motor : unitree_left->motors) {
            unitree_msgs::msg::MotorState motor_msg;
            motor_msg.q = static_cast<float>(motor.data.q);
            motor_msg.dq = static_cast<float>(motor.data.dq);
            motor_msg.temperature = static_cast<int>(motor.data.temp);
            int id = motor.id;
            msg_list.motor_state[id] = motor_msg;
            jointstate_msg.position[id] = static_cast<float>(motor.data.q);
            jointstate_msg.velocity[id] = static_cast<float>(motor.data.dq);
        }

        // Process motors from the second unitree interface
        for (const auto &motor : unitree_right->motors) {
            unitree_msgs::msg::MotorState motor_msg;
            motor_msg.q = static_cast<float>(motor.data.q);
            motor_msg.dq = static_cast<float>(motor.data.dq);
            motor_msg.temperature = static_cast<int>(motor.data.temp);

            int id = motor.id;
            msg_list.motor_state[id] = motor_msg;
            jointstate_msg.position[id] = static_cast<float>(motor.data.q);
            jointstate_msg.velocity[id] = static_cast<float>(motor.data.dq);
        }

        // Publish the messages
        status_pub->publish(msg_list);
        jointstate_pub->publish(jointstate_msg);
    }
    rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char *argv[]) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<UnitreeInterface>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: crazydog_ws/src/robot_interfaces/robot_interfaces/unitree_pubsub.py
import time
import math
import sys
sys.path.append('/home/crazydog/crazydog/crazydog_ws/src/robot_interfaces/robot_interfaces/unitree_actuator_sdk/lib')
sys.path.append('..')
# sys.path.append('/home/crazydog/crazydog/crazydog_ws/src/lqr_control/lqr_control')
from unitree_actuator_sdk import * # type: ignorei
import threading

import rclpy
from rclpy.node import Node
from unitree_msgs.msg import LowCommand, LowState, MotorCommand, MotorState
from std_msgs.msg import Float32MultiArray
from sensor_msgs.msg import JointState

MOTOR_INIT_POS = [None, 0.669, 1.1, None, 7.5302, 2.2]

MOTOR_ORIGIN_POS = [0.0, -4.6, 27.8, 0.0, 12.8, -24.4, 0.0, 0.0]
SCALE = [6.33, 6.33, 6.33*1.6, 6.33, -6.33, -6.33*1.6, 1.0, 1.0]
WHEEL_RADIUS = 0.07     # m

class unitree_communication(object):
    def __init__(self,device_name = '/dev/ttyUSB0'):
        self.serial = SerialPort(device_name)
        self.motors = []
        # self.runing_flag = False

    def createMotor(self, motor_number=None, MAX=None, MIN=None):
        if motor_number not in [motor.id for motor in self.motors]:
            motor = unitree_motor(motor_number, MAX_degree=MAX, MIN_degree=MIN)
            self.motors.append(motor)
            return motor                              

        else:
            print("Motor {0} already exist".format(motor_number))
            for motor in self.motors:
                if motor.cmd.id == motor_number:
                    return motor

    def position_force_velocity_cmd(self,motor_number=0,torque=0,kp=0,kd=0,position=0,velocity=0):
        for motor in self.motors:
            if motor.id == motor_number:
                motor.cmd.mode = queryMotorMode(MotorType.GO_M8010_6,MotorMode.FOC)
                motor.cmd.tau = torque
                motor.cmd.kp = kp
                motor.cmd.kd = kd
                motor.cmd.q = position  
                motor.cmd.dq = velocity*queryGearRatio(MotorType.GO_M8010_6)
            
    def motor_sendRecv(self):
        success = True
        for motor in self.motors:
            if motor.min <= motor.data.q <= motor.max:
                self.serial.sendRecv(motor.cmd, motor.data)
                # time.sleep(0.1)
            else:
                print("motor {0} out off constrant".format(motor.cmd.id))
                print(motor.max, motor.data.q, motor.min)
                motor.cmd.mode = queryMotorMode(MotorType.GO_M8010_6,MotorMode.FOC)
                motor.cmd.q    = 0
                motor.cmd.dq   = 0
                motor.cmd.kp   = 0
                motor.cmd.kd   = 0
                motor.cmd.tau  = 0
                self.serial.sendRecv(motor.cmd, motor.data)
                success = False
                # time.sleep(0.0006)
        return success

    def enableallmotor(self):       
        for motor in self.motors:
            motor.cmd.mode = queryMotorMode(MotorType.GO_M8010_6,MotorMode.FOC)
            motor.cmd.q    = 0
            motor.cmd.dq   = 0
            motor.cmd.kp   = 0
            motor.cmd.kd   = 0
            motor.cmd.tau  = 0
            self.serial.sendRecv(motor.cmd, motor.data)
            time.sleep(0.01)
            print(motor.cmd.id,motor.data.q)

    
class unitree_motor(object):                                                                                  
    def __init__(self, motor_id=None,MAX_degree=None,MIN_degree=None):
        self.id = motor_id
        self.cmd = MotorCmd()
        self.data = MotorData()
        self.data.motorType = MotorType.GO_M8010_6
        self.cmd.motorType = MotorType.GO_M8010_6
        self.max = MAX_degree * SCALE[motor_id] + MOTOR_ORIGIN_POS[motor_id]
        self.min = MIN_degree * SCALE[motor_id] + MOTOR_ORIGIN_POS[motor_id]
        self.cmd.id = motor_id

        print(f'constrain: id {self.id}, max {self.max}, min {self.min}')


class UnitreeInterface(Node):

    def __init__(self):
        super().__init__('unitree_pubsub')

        self.unitree = unitree_communication('/dev/unitree-l')
        MOTOR1 = self.unitree.createMotor(motor_number = 1, MAX=2.093, MIN=0.0)
        MOTOR2 = self.unitree.createMotor(motor_number = 2, MAX=0.0, MIN=-2.7)
        self.unitree2 = unitree_communication('/dev/unitree-r')
        MOTOR4 = self.unitree2.createMotor(motor_number = 4, MAX=0.0, MIN=2.093)
        MOTOR5 = self.unitree2.createMotor(motor_number = 5, MAX=-2.7, MIN=0.0)

        self.unitree_command_sub = self.create_subscription(
            LowCommand,
            'unitree_command',
            self.command_callback,
            1)
        self.status_pub = self.create_publisher(LowState, 'unitree_status', 1)
        self.unitree.enableallmotor()
        self.unitree2.enableallmotor()
        self.recv_timer = self.create_timer(0.001, self.recv_timer_callback)    # period need to be check

        self.jointstate_pub = self.create_publisher(JointState, 'jointstate', 1)
        self.foc_status_sub = self.create_subscription(
            Float32MultiArray,
            'foc_msg',
            self.foc_status_callback,
            1)
        self.jointstate_msg = JointState()
        self.jointstate_msg.header.stamp = self.get_clock().now().to_msg()
        self.jointstate_msg.header.frame_id = ""
        self.jointstate_msg.name = ["hip_l", "thigh_l","calf_l","hip_r","thigh_r","calf_r", "wheel_l", "wheel_r"]
        self.jointstate_msg.position = [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
        self.jointstate_msg.velocity = [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]

    def command_callback(self, msg):
        for id, cmd in enumerate(msg.motor_cmd):
            motor_number = id
            torque = cmd.tau
            kp = cmd.kp
            kd = cmd.kd
            position = cmd.q
            velocity = cmd.dq
            self.unitree.position_force_velocity_cmd(motor_number, torque, kp, kd, position, velocity)
            self.unitree2.position_force_velocity_cmd(motor_number, torque, kp, kd, position, velocity)

    def foc_status_callback(self, msg: Float32MultiArray):
        self.jointstate_msg.header.stamp = self.get_clock().now().to_msg()
        if msg.data[0] == 513.:   # motor left
            self.jointstate_msg.position[6] = -msg.data[1]
            self.jointstate_msg.velocity[6] = -msg.data[2] * WHEEL_RADIUS * (2 * math.pi / 60)
        elif msg.data[0] == 514.: # motor right
            self.jointstate_msg.position[7] = msg.data[1]
            self.jointstate_msg.velocity[7] = msg.data[2] * WHEEL_RADIUS * (2 * math.pi / 60)
        self.jointstate_pub.publish(self.jointstate_msg)

    def recv_timer_callback(self):
        feedback = self.unitree.motor_sendRecv()
        feedback2 = self.unitree2.motor_sendRecv()
        if feedback==False or feedback2==False:
            self.get_logger().error('unitree motor out of constrain.')
        msg_list = LowState()
        self.jointstate_msg.header.stamp = self.get_clock().now().to_msg()
        
        for motor in self.unitree.motors:
            msg = MotorState()
            msg.q = float(motor.data.q)
            msg.dq = float(motor.data.dq)
            msg.temperature = int(motor.data.temp)
            id = motor.id
            msg_list.motor_state[id] = msg
            self.jointstate_msg.position[id] = float(motor.data.q)
            self.jointstate_msg.velocity[id] = float(motor.data.dq)
        for motor in self.unitree2.motors:
            msg = MotorState()
            msg.q = float(motor.data.q)
            msg.dq = float(motor.data.dq)
            msg.temperature = int(motor.data.temp)
            id = motor.id
            msg_list.motor_state[id] = msg
            self.jointstate_msg.position[id] = float(motor.data.q)
            self.jointstate_msg.velocity[id] = float(motor.data.dq)
        self.status_pub.publish(msg_list)
        self.jointstate_msg = self.scaling(self.jointstate_msg)
        self.jointstate_pub.publish(self.jointstate_msg)
    
    def scaling(self, states: JointState):
        states.position = [(state-org)/scale for state, scale, org in zip(states.position, SCALE, MOTOR_ORIGIN_POS)]
        states.velocity = [state/scale for state, scale in zip(states.velocity, SCALE)]
        return states
    
def main(args=None):
    rclpy.init(args=args)

    unitree_interface = UnitreeInterface()

    rclpy.spin(unitree_interface)

    # Destroy the node explicitly
    # (optional - otherwise it will be done automatically
    # when the garbage collector destroys the node object)
    unitree_interface.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()